### PR TITLE
raftstore: trigger compaction when no valid split key can be found

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #15282

What's Changed:

```commit-message
trigger compaction when no valid split key can be found
```
This PR handles the case where split keys cannot be found due to so many duplicated mvcc keys.
It triggers range compaction for this case but at most once for each safe point duration to avoid scheduling to many compaction tasks.

The test steps are:
1. prepare few rows, say 20 rows
2. update these 20 rows and regions will be split to 20 regions and no extra region will be splitted.
3. before safe point is updated, stop the workload, so no minor compaction will be performed
4. use pd to trigger split a region
5. region size decreases significantly
image

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -4090,7 +4098,70 @@ where
...
+                // trigger a compaction to handle it.
+                if e.to_string().contains(NO_VALID_SPLIT_KEY) {
+                    let safe_ts = (|| {
+                        fail::fail_point!("safe_point_inject", |t| {
+                            t.unwrap().parse::<u64>().unwrap()
+                        });
+                        poll_ctx.safe_point.load(Ordering::Relaxed)
+                    })();
+                    if safe_ts <= self.last_record_safe_point {
+                        info!(
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/peer.rs:4118]: Will this log be too verbose? can it be debug

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -4090,7 +4098,70 @@ where
...
+                            "skip schedule compact range due to safe_point not updated";
+                            "region_id" => self.region_id,
+                            "safe_point" => safe_ts,
+                        );
+                        return e;
+                    }
+                    self.last_record_safe_point = safe_ts;
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/peer.rs:4125]: it should only be updated when all tasks are scheduled

**Review by tonyxuqqi — APPROVED** [review]
Minor comments. Overall LGTM.

File: components/raftstore/src/store/peer.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/peer.rs:4118]: Good suggestion.

File: components/raftstore/src/store/peer.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/peer.rs:4125]: Good catch.

**Review by bufferflies — APPROVED** [review]

**Pull request merged**